bika/lims/tests/doctests/ARAnalysesField.rst

```diff
@@ -0,0 +1,603 @@
+    >>> new_analyses = field.set(ar, [analysis1], specs=[ph_specs])
+
+We expect to have now just one Analysis set:
+
+    >>> analyses = field.get(ar, full_objects=True)
+    >>> analyses
+    [<Analysis at /plone/clients/client-1/water-0001-R01/PH>]
+
+And the specification should be according to the values we have set
+
+    >>> ph = analyses[0]
+    >>> phspec = ph.getResultsRange()
+
+    >>> phspec.get("min")
+    5.2
+
+    >>> phspec.get("max")
+    7.9
+
+    >>> phspec.get("error")
+    3
+
+
+Setting Analyses Prices
+.......................
+
+Prices are primarily defined on Analyses Services:
+
+    >>> analysisservice1.getPrice()
+    '10.00'
+
+    >>> analysisservice2.getPrice()
+    '20.00'
+
+    >>> analysisservice3.getPrice()
+    '30.00'
+
+Created Analyses inherit that price:
+
+    >>> new_analyses = field.set(ar, all_services)
+
+    >>> analysis1 = ar[analysisservice1.getKeyword()]
+    >>> analysis2 = ar[analysisservice2.getKeyword()]
+    >>> analysis3 = ar[analysisservice3.getKeyword()]
+
+    >>> analysis1.getPrice()
+    '10.00'
+
+    >>> analysis2.getPrice()
+    '20.00'
+
+    >>> analysis3.getPrice()
+    '30.00'
+
+The `setter` also allows to set custom prices for the Analyses:
+
+    >>> prices = {
+    ...     analysisservice1.UID(): "100",
+    ...     analysisservice2.UID(): "200",
+    ...     analysisservice3.UID(): "300",
+    ... }
+
+Now we set the field with all analyses services and new prices:
+
+    >>> new_analyses = field.set(ar, all_services, prices=prices)
+
+The Analyses have now the new prices:
+
+    >>> analysis1.getPrice()
+    '100.00'
+
+    >>> analysis2.getPrice()
+    '200.00'
+
+    >>> analysis3.getPrice()
+    '300.00'
+
+The Services should retain the old prices:
+
+    >>> analysisservice1.getPrice()
+    '10.00'
+
+    >>> analysisservice2.getPrice()
+    '20.00'
+
+    >>> analysisservice3.getPrice()
+    '30.00'
+
+
+Calculations and Interim Fields
+...............................
+
+When an Analysis is assigned to an AR, it inherits its Calculation and Interim Fields.
+
+Create some interim fields:
+
+    >>> interim1 = {"keyword": "A", "title": "Interim A", "value": 1, "hidden": False, "type": "int", "unit": "x"}
+    >>> interim2 = {"keyword": "B", "title": "Interim B", "value": 2, "hidden": False, "type": "int", "unit": "x"}
+    >>> interim3 = {"keyword": "C", "title": "Interim C", "value": 3, "hidden": False, "type": "int", "unit": "x"}
+    >>> interim4 = {"keyword": "D", "title": "Interim D", "value": 4, "hidden": False, "type": "int", "unit": "x"}
+
+Append interim field `A` to the `Total Hardness` Calculation:
+
+    >>> calc4.setInterimFields([interim1])
+    >>> map(lambda x: x["keyword"], calc4.getInterimFields())
+    ['A']
+
+Append interim field `B` to the `Total Hardness` Analysis Service:
+
+    >>> analysisservice4.setInterimFields([interim2])
+    >>> map(lambda x: x["keyword"], analysisservice4.getInterimFields())
+    ['B']
```

Comment:
If you want to omit the tricky workaround in `ARAnalysisField` regarding interim (something I agree), then this  should be like follows:

```python
    >>> map(lambda x: x["keyword"], analysisservice4.getInterimFields())
    ['A', 'B']
```
Remember that at AS level, you can override the default value for an Interim defined in the Calculation, so you could have an interim called "Ca" in calculation with a default value of 0.1 and with the attribute "Hidden" checked whilst you could have an interim with the same keyword ("Ca"), but with default value 0.5 and "Hidden=False". So, InterimFields set at AS level have always priority over the interims set at Calculation level.

If you add an interim to a Calculation, `InterimFields` field from all Analysis Services that have that Calculation assigned, must be updated accordingly. New interims that were not present in `AnalysisService` must be added accordingly (otherwise, the system will not be able to calculate the result). Those interims removed from `Calculation` must be removed from the `AnalysisService`. Those interims that were added directly in the Analysis Service that do not match with the interims set in `Calcultion` must be preserved.

Reply:
You could argue that well, leaving this like this, the system still works. Yes, agree, but because we have to always take this priority thing with interims into consideration (so at some point, it works by accident). See this as an example:

https://github.com/senaite/senaite.core/blob/master/bika/lims/browser/calcs.py#L190-L205

First, we get the interim fields from the `Calculation` and just after mapping them, we override with the values from the analysis.... Wouldn't it be better to just get rid of `calc.getInterimFields()`and reduce complexity? The code would look much better and clean, with just only one loop.